### PR TITLE
Gives WorkflowAction access to props.

### LINKED ISSUE #56: Pass props to WorkflowActions
It's come up a couple of times, at least in Kotlin, that `WorkflowAction`s get the state of the current workflow but not its props. Since we consider props to be part of the workflow's overall "state", it seems unintuitive that we only pass part of that to actions.

The reason for _not_ doing this was, initially, to discourage side effects from being performed in actions. However, we realized that is too restrictive (see [this Slack discussion](https://workflow-community.slack.com/archives/CHB9VKA10/p1572625798145200)).

_Note for Swift this means passing the `Workflow` value itself into the action, since there's no separate props type._

_Kotlin version of square/workflow#919._

### PR DESCRIPTION
**Pull request by rjrjr**
Closes #56.

### CODE REVIEW

File: workflow-core/src/main/java/com/squareup/workflow/WorkflowAction.kt [hunk 1]
```diff
@@ -73,12 +76,12 @@ interface WorkflowAction<StateT, out OutputT> {
...
+    fun <PropsT, StateT, OutputT> noAction(): WorkflowAction<PropsT, StateT, OutputT> =
+      NO_ACTION as WorkflowAction<Any?, StateT, OutputT>
 
     /**
-     * Convenience function that returns a [WorkflowAction] that will just set the state to [newState]
-     * (without considering the current state) and optionally emit an output.
+     * Convenience function that returns a [WorkflowAction] that will just set the state to
+     * [newState] (without considering the current state) and optionally emit an output.
      */
     @Deprecated(
```
Comment (rjrjr) [inline workflow-core/src/main/java/com/squareup/workflow/WorkflowAction.kt:86]: @zach-klippenstein  We should delete these, they're so dangerous. Is that tracked?

Comment (zach-klippenstein) [reply workflow-core/src/main/java/com/squareup/workflow/WorkflowAction.kt:86]: I don't think so. We don't have any issues for removing deprecated APIs afaik.

Comment (rjrjr) [reply workflow-core/src/main/java/com/squareup/workflow/WorkflowAction.kt:86]: These are the only ones I really care about. After this merges I'll put up a PR to snip them. Not worth tracking.

Comment (zach-klippenstein) [reply workflow-core/src/main/java/com/squareup/workflow/WorkflowAction.kt:86]: I thought they were still extremely popular.

Comment (rjrjr) [reply workflow-core/src/main/java/com/squareup/workflow/WorkflowAction.kt:86]: They are, but they're also often extremely a bad idea, and misused more often than not. And easy enough to recreate on your own w/o us sanctioning them.

**Comment by rjrjr** [comment]
> The only reason that the render context needs to be aware of the props type is so that type inference can infer the type of actions when sent to action sinks and stuff, right?

And for the actions tied to any worker or workflow output.

File: workflow-core/src/main/java/com/squareup/workflow/WorkflowAction.kt (hunk 1 above)
Comment (rjrjr) [reply workflow-core/src/main/java/com/squareup/workflow/WorkflowAction.kt:86]: https://github.com/square/workflow-kotlin/pull/85

**Pull request merged**